File: frontend/lib/itsm/priority.ts
import type { Priority, PriorityMatrix } from "./types";

/** The standard ITIL Priority Matrix — matrix[impact][urgency] -> priority.
 *  Mirrors the server default (apps.itsm_projects.models.default_priority_matrix);
 *  used as a fallback when a project hasn't stored one. */
export const DEFAULT_PRIORITY_MATRIX: PriorityMatrix = {
  critical: { high: "critical", medium: "critical", low: "high" },
  high: { high: "critical", medium: "high", low: "medium" },
  medium: { high: "high", medium: "medium", low: "low" },
  low: { high: "medium", medium: "low", low: "low" },
};

export const IMPACT_OPTIONS: { value: string; label: string }[] = [
  { value: "low", label: "Low" },
  { value: "medium", label: "Medium" },
  { value: "high", label: "High" },
  { value: "critical", label: "Critical" },
];

export const URGENCY_OPTIONS: { value: string; label: string }[] = [
  { value: "low", label: "Low" },
  { value: "medium", label: "Medium" },
  { value: "high", label: "High" },
];

/** Derive Priority from Impact × Urgency using the project's matrix (or the default).
 *  Returns `null` when either input is blank or the matrix has no cell — the caller
 *  keeps the current priority in that case. Mirrors
 *  apps.itsm_tickets.services.priority.compute_priority. */
export function computePriority(
  matrix: PriorityMatrix | undefined | null,
  impact: string | undefined | null,
  urgency: string | undefined | null,
): Priority | null {
  if (!impact || !urgency) return null;
  const m = matrix || DEFAULT_PRIORITY_MATRIX;
  const row = m[impact];
  if (!row) return null;
  return (row[urgency] as Priority) || null;
}

File: backend/apps/itsm_tickets/services/priority.py
"""ITIL Priority Matrix — the single site that derives Priority from Impact × Urgency.

The matrix lives on the project (``Project.priority_matrix``, per-project customisable,
defaulting to the standard ITIL matrix). ``compute_priority`` is the authoritative
server-side calculation; the frontend mirrors it (``lib/itsm/priority.ts``) for the live
UI preview. Keep the two in sync.
"""

from __future__ import annotations


def compute_priority(project, impact, urgency):
    """Return the priority code for ``impact × urgency`` per ``project``'s matrix.

    Returns ``None`` when either input is blank or the matrix has no cell for the
    pair — callers keep the ticket's current priority in that case (so a partial
    assessment never blanks or wrongly downgrades priority). Falls back to the
    standard ITIL matrix when the project has no matrix stored.
    """
    from apps.itsm_projects.models import default_priority_matrix

    if not impact or not urgency:
        return None
    matrix = (project.priority_matrix if project is not None else None) or default_priority_matrix()
    row = matrix.get(impact)
    if not isinstance(row, dict):
        return None
    return row.get(urgency) or None
